### PR TITLE
feat(report): support to query policy from remote OPA server

### LINKED ISSUE #1785: Trivy config to query policy from an OPA server
Can trivy config scanner query OPA policies from an OPA server rather than loading from an OPA bundle ?

### PR DESCRIPTION
**Pull request by NitroCao**
## Description
Add support to query policy from remote OPA server when filtering vulnerabilities.

Implementation:
* integrate with the REST API of OPA. Related documentation is here: [Integrating with the REST API](https://www.openpolicyagent.org/docs/latest/integration/#integrating-with-the-rest-api)

* I reused the `--ignore-policy` option. If the value is a valid URL, we'll query from the server. Otherwise, we'll treat it as a local policy file. But I don't think it's a good idea. Advise please.

* I preserved the built-in library `lib.rego` if the value of `--ignore-policy` is a local policy file. But I think it's redundant, we can migrate it to the documentation.

* As for the query, I still reused the previous `data.trivy.ignore`.

## Related issues
- Close #1785 

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [ ] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by knqyf263** [comment]
Thanks for your contribution! Looks really promising! We'll be on holidays next week, but hope that we'll review it soon.

File: pkg/result/policy.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+type PolicyStore interface {
+	Evaluate(ctx context.Context, input interface{}) (bool, error)
+}
+
+type LocalPolicyStore struct {
+	query rego.PreparedEvalQuery
+}
+
+func NewLocalPolicyStore(ctx context.Context, policyFile string) (PolicyStore, error) {
+	policy, err := ioutil.ReadFile(policyFile)
```
Comment (knqyf263) [inline pkg/result/policy.go:26]: `ioutil` is deprecated. Please use os.ReadFile.
https://pkg.go.dev/os#ReadFile

File: pkg/result/policy.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	if err != nil {
+		err = xerrors.Errorf("unable to read policy file %s: %w", policyFile, err)
+		return nil, err
```
Comment (knqyf263) [inline pkg/result/policy.go:29]:
```suggestion
		return nil, xerrors.Errorf("unable to read policy file %s: %w", policyFile, err)
```

File: pkg/result/policy.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+type RemotePolicyStore struct {
+	remoteURL string
+}
+
+func NewRemotePolicyStore(remoteAddr string) (PolicyStore, error) {
+	u, err := url.Parse(remoteAddr)
+	if err != nil {
+		return nil, err
```
Comment (knqyf263) [inline pkg/result/policy.go:67]: Could you wrap the error?

File: pkg/result/policy.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	}
+	u.Path = path.Join(u.Path, "/v1/data/trivy/ignore")
```
Comment (knqyf263) [inline pkg/result/policy.go:69]: What if we define the path as a constant?

File: pkg/result/policy.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+		remoteURL: u.String(),
+	}, nil
+}
+
+func (r *RemotePolicyStore) Evaluate(ctx context.Context, input interface{}) (bool, error) {
+	reqBody, err := processQueryInput(input)
+	if err != nil {
+		return false, xerrors.Errorf("unable to process policy input: %w", err)
+	}
+	req, err := http.NewRequestWithContext(ctx, "POST", r.remoteURL, reqBody)
```
Comment (knqyf263) [inline pkg/result/policy.go:81]:
```suggestion
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, r.remoteURL, reqBody)
```

File: pkg/result/policy.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+func processQueryInput(input interface{}) (io.Reader, error) {
+	type opaRequest struct {
+		Input interface{} `json:"input"`
+	}
+
+	inputData := opaRequest{Input: input}
+	inputBytes, err := json.Marshal(&inputData)
+	if err != nil {
+		err = xerrors.Errorf("unable to marshal input data: %w", err)
+		return nil, err
```
Comment (knqyf263) [inline pkg/result/policy.go:106]:
```suggestion
		return nil, xerrors.Errorf("unable to marshal input data: %w", err)
```

File: pkg/result/policy.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+		return false, err
+	}
+	if respBytes, err = ioutil.ReadAll(resp.Body); err != nil {
+		err = xerrors.Errorf("unable to read query response: %w", err)
+		return false, err
+	}
+	if err = json.Unmarshal(respBytes, &queryResult); err != nil {
+		err = xerrors.Errorf("unable to unmarshal query response: %w", err)
+		return false, err
+	}
```
Comment (knqyf263) [inline pkg/result/policy.go:137]:
```suggestion
	if err = json.NewDecoder(resp.Body).Decode(&queryResult); err != nil {
		return false, xerrors.Errorf("unable to unmarshal query response: %w", err)
	}
```

File: pkg/result/policy.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+	var queryResult opaResult
+	var err error
+	var respBytes []byte
+
+	// From the API doc of OPA, non-HTTP 200 response codes indicate configuration
+	// or runtime errors.
+	if resp.StatusCode != http.StatusOK {
+		err = xerrors.Errorf("unable to get query result, response code is %d, want 200 instead", resp.StatusCode)
+		return false, err
```
Comment (knqyf263) [inline pkg/result/policy.go:128]:
```suggestion
		return false, xerrors.Errorf("unable to get query result, response code is %d, want 200 instead", resp.StatusCode)
```

File: pkg/result/policy.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+		err = xerrors.Errorf("unable to create new policy request: %w", err)
+		return false, err
+	}
+	req.Header.Add("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		err = xerrors.Errorf("unable to send query policy request: %w", err)
+		return false, err
+	}
```
Comment (knqyf263) [inline pkg/result/policy.go:92]: Missing Close()?

```suggestion
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		err = xerrors.Errorf("unable to send query policy request: %w", err)
		return false, err
	}
	defer resp.Close()
```

File: pkg/result/policy_test.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+`),
+		"bad.rego": []byte(`
+package trivy
+
+invalid rego file
+`),
+	}
+	tempDir := t.TempDir()
+	for filename, content := range policyFiles {
+		if err := ioutil.WriteFile(filepath.Join(tempDir, filename), content, 0600); err != nil {
```
Comment (knqyf263) [inline pkg/result/policy_test.go:35]: `ioutil` is deprecated

File: pkg/result/policy_test.go [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+			t.Fatalf("unable to create temp file %s: %s", filename, err)
+		}
```
Comment (knqyf263) [inline pkg/result/policy_test.go:37]:
```suggestion
		err := ioutil.WriteFile(filepath.Join(tempDir, filename), content, 0600)
		require.NoError(t, err)
```

File: pkg/result/result.go [hunk 12]
```diff
@@ -228,24 +228,22 @@ func summarize(status types.MisconfStatus, summary *types.MisconfSummary) {
...
+	var err error
+	var policyStore PolicyStore
 
-	query, err := rego.New(
-		rego.Query("data.trivy.ignore"),
-		rego.Module("lib.rego", module),
-		rego.Module("trivy.rego", string(policy)),
-	).PrepareForEval(ctx)
+	if utils.IsValidURL(policyFile) {
+		policyStore, err = NewRemotePolicyStore(policyFile)
```
Comment (knqyf263) [inline pkg/result/result.go:235]: Can we add a test case for RemotPolicyStore in result_test.go? I think we can use httptest.

**Review by knqyf263 — COMMENTED** [review]
Looks good. I left some small comments. Also, could you update documentation?

**Comment by NitroCao** [comment]
Resolved all requested changes. Hope I didn't miss something

**Comment by NitroCao** [comment]
@knqyf263 Is there any suggestion? I have enough time theses days and hope we can finish this PR as soon as possible 😀

**Pull request closed without merging**